pkg/reconciler/eventtype/eventtype.go

```diff
@@ -0,0 +1,208 @@
+	Type        string
+	Source      string
+	Schema      string
+	Broker      string
+	Description string
+}
+
+// syncEventTypes is a helper struct used to sync EventTypes during the reconciliation process.
+type syncEventTypes struct {
+	ToCreate []eventingv1alpha1.EventType
+	ToDelete []eventingv1alpha1.EventType
+}
+
+// ReconcileEventTypes reconciles the EventTypes taken from 'args', and sets 'owner' as the controller.
+func (r *Reconciler) ReconcileEventTypes(ctx context.Context, owner metav1.Object, args *ReconcilerArgs) error {
+	current, err := r.getEventTypes(ctx, args.Namespace, args.Labels, owner)
+	if err != nil {
+		return err
+	}
+
+	expected, err := r.makeEventTypes(args, owner)
+	if err != nil {
+		return err
+	}
+	sync := r.getEventTypesToSync(current, expected)
+	for _, eventType := range sync.ToCreate {
+		err = r.Client.Create(ctx, &eventType)
+		if err != nil {
+			return err
+		}
+	}
+	// Deletion might not be needed, but just to make sure we dedupe EventTypes
+	// in case some source controller has a bug.
+	for _, eventType := range sync.ToDelete {
+		err = r.Client.Delete(ctx, &eventType)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// getEventTypes returns the EventTypes controlled by 'owner'.
+func (r *Reconciler) getEventTypes(ctx context.Context, namespace string, lbs map[string]string, owner metav1.Object) ([]eventingv1alpha1.EventType, error) {
+	eventTypes := make([]eventingv1alpha1.EventType, 0)
+
+	opts := &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: labels.SelectorFromSet(lbs),
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+	for {
+		el := &eventingv1alpha1.EventTypeList{}
+		if err := r.Client.List(ctx, opts, el); err != nil {
+			return nil, err
+		}
+
+		for _, e := range el.Items {
+			if metav1.IsControlledBy(&e, owner) {
+				eventTypes = append(eventTypes, e)
+			}
+		}
+		if el.Continue != "" {
+			opts.Raw.Continue = el.Continue
+		} else {
+			return eventTypes, nil
+		}
+	}
+}
+
+// makeEventTypes creates the in-memory representation of the EventTypes.
+func (r *Reconciler) makeEventTypes(args *ReconcilerArgs, owner metav1.Object) ([]eventingv1alpha1.EventType, error) {
+	eventTypes := make([]eventingv1alpha1.EventType, 0)
+	for _, arg := range args.EventTypes {
+		eventType := r.makeEventType(arg, args.Namespace, args.Labels)
+		// Setting the reference to delete the EventType upon uninstalling the source.
+		if err := controllerutil.SetControllerReference(owner, &eventType, r.Scheme); err != nil {
+			return nil, err
+		}
+		eventTypes = append(eventTypes, eventType)
+	}
+	return eventTypes, nil
+}
+
+// makeEventType creates the in-memory representation of the EventType.
+func (r *Reconciler) makeEventType(arg *EventTypeArgs, namespace string, lbl map[string]string) eventingv1alpha1.EventType {
```

Comment:
_This_ should be in the resource dir for this reconciler.

Reply:
this is not technically a reconciler but rather a helper not to duplicate code across sources controllers.
But yes, I can move it there...

Reply:
Actually, do you think this file should go inside pkg/reconciler/eventtype? 
Or just leave it in pkg/reconciler? as a sort of helper for all other reconcilers, as this is technically not a reconciler.
Or maybe just put it outside, pkg/eventtype?

Reply:
IMO this whole thing is effectively its own reconciler, and should have its own `/reconciler` directory with a `/resources` subdir. But again I don't know how elegant that will be in practice.

Reply:
Will address that structure change in the follow up once we go with your approach to reduce boilerplate code.